games/xworld3d/x3item.cpp

```diff
@@ -92,12 +123,33 @@ void X3Agent::jump() {
     }
 }
 
-float X3Agent::reach_test(const Pose& gpose) {
-    float dx = gpose.x() - pose().x();
-    float dy = gpose.y() - pose().y();
-    float dz = gpose.z() - pose().z();
-    float d = sqrt(dx * dx + dy * dy);
-    float reaching_score = -1; // lower end of cos range
+X3ItemPtr X3Agent::collect_item(const std::map<std::string, X3ItemPtr>& items,
+                                const std::string& type) {
+    X3ItemPtr item = nullptr;
+    set_speed(0.0f, 0.0f, 0.0f);
+    // the angle between the agent's facing direction and
+    // the direction from the agent to the item should be less than 45 degrees
+    x3real best_score = 0.707; // 45 degrees is the minimum
+    x3real score;
+    for (auto& kv : items) {
+        if (kv.second->type() == type) {
+            score = reach_test(kv.second->pose());
+            if (score > best_score) {
+                item = kv.second;
+                best_score = score;
+            }
+        }
+    }
+    return item;
+}
+
+x3real X3Agent::reach_test(const Pose& pose) {
+    const Pose self = object_.pose();
```

Comment:
Needs a better name. This is actually not testing reach but testing facing direction